### PR TITLE
Deprecate old Sort functions and replace with new argument lists

### LINKED ISSUE #25544: Deprecate old `Sort` functions and replace with new functions
The new sort module API has been designed in https://github.com/chapel-lang/chapel/issues/24857, this is an issue describing an implementation task

Old functions to deprecate
```chapel
proc sort(ref Data: [?Dom] ?eltType, comparator: ?rec = defaultComparator, param stable: bool = false, param inPlaceAlgorithm: bool = false)
proc isSorted(Data: [?Dom] ?eltType, comparator: ?rec = defaultComparator) : bool
iter sorted(x, comparator: ?rec = defaultComparator)
```

New functions to add
```chapel
proc sort(ref x: [], comparator: ? = new defaultComparator(), param stable: bool = false)
proc isSorted(x: [], comparator: ? = new defaultComparator()): bool
iter sorted(x, comparator: ? = new defaultComparator())
```

### PR DESCRIPTION
**Pull request by lydia-duncan**
Resolves #25544 

Implements the decisions reached by the Sort discussion team.  Updates the argument lists for:
- `sort` by renaming the array argument and removing the `inPlaceAlgorithm` argument (which was confusing), and removing the queries.  Also uses `new DefaultComparator()` instead of the module-scoped variable.
- `isSorted` by renaming the array argument, removing the queries and using `new DefaultComparator()` instead of the module-scoped variable.
- and the `sorted` iterator by removing the queries and using `new DefaultComparator()` instead of the module-scoped variable.

The sorted iterator did not need a deprecation for this change.

Adds tests locking in the deprecation messages for `sort` and `isSorted`.

Passed a full paratest with futures

### CODE REVIEW

File: modules/standard/Sort.chpl [hunk 1]
```diff
@@ -471,6 +551,8 @@ The choice of sorting algorithm used is made by the implementation.
   sort. If it is ``true``, it will use an in-place algorithm in order to use
   less memory.
  */
+pragma "last resort"
+@deprecated("The 'sort' function with 'Data' and 'inPlaceAlgorithm' arguments has been deprecated, please use the 'sort' function with an 'x' argument instead")
 proc sort(ref Data: [?Dom] ?eltType, comparator:?rec=defaultComparator,
```
Comment (jabraham17) [inline modules/standard/Sort.chpl:556]: If `inPlaceAlgorithm=false`, can this call the new function? that way the logic isn't duplicated

File: modules/standard/Sort.chpl [hunk 2]
```diff
@@ -545,13 +626,39 @@ proc isSorted(Data: [?Dom] ?eltType, comparator:?rec=defaultComparator): bool {
...
+@deprecated("'isSorted' with the argument name 'Data' is deprecated, please use the version with the argument name 'x' instead")
+proc isSorted(Data: [?Dom] ?eltType, comparator:?rec=defaultComparator): bool {
+  return isSorted(x=Data, comparator);
+}
+
+
+pragma "last resort"
+@chpldoc.nodoc
+/* Error message for multi-dimension arrays */
+@deprecated("'isSorted' with the argument name 'Data' is deprecated, please use the version with the argument name 'x' instead")
```
Comment (jabraham17) [inline modules/standard/Sort.chpl:654]: Does the error overload need a deprecation?

File: modules/standard/Sort.chpl [hunk 3]
```diff
@@ -545,13 +626,39 @@ proc isSorted(Data: [?Dom] ?eltType, comparator:?rec=defaultComparator): bool {
 
 @chpldoc.nodoc
 /* Error message for multi-dimension arrays */
+proc isSorted(x: [], comparator:? =new DefaultComparator())
```
Comment (jabraham17) [inline modules/standard/Sort.chpl:629]: nit, spacing on `=new`

File: modules/standard/Sort.chpl [hunk 4]
```diff
@@ -518,25 +600,24 @@ proc sort(ref Data: [?Dom] ?eltType, comparator:?rec=defaultComparator)
...
+   :arg x: The array to verify
+   :type x: `array`
    :arg comparator: :ref:`Comparator <comparators>` record that defines how the
       data is sorted.
    :returns: ``true`` if array is sorted
    :rtype: `bool`
  */
-proc isSorted(Data: [?Dom] ?eltType, comparator:?rec=defaultComparator): bool {
-  chpl_check_comparator(comparator, eltType);
+proc isSorted(x: [], comparator:? = new DefaultComparator()): bool {
```
Comment (jabraham17) [inline modules/standard/Sort.chpl:613]: I think this probably should have `const D = x.domain` (or whatever name you prefer) at the top of the function to manually LICM the domain query

File: modules/standard/Sort.chpl [hunk 5]
```diff
@@ -423,6 +423,86 @@ proc radixSortOk(Data: [?Dom] ?eltType, comparator) param {
...
+  data is sorted.
+:arg stable: Defaults to ``false``. If it is ``false``, the implementation
+  can sort in a way that reorders equal keys. If it is ``true``, it will use a
+  stable algorithm in order to preserve the order of equal keys.
+ */
+proc sort(ref x: [], comparator:? = new DefaultComparator(),
+          param stable:bool = false) {
+  chpl_check_comparator(comparator, x.eltType);
+
+  if x.domain.low >= x.domain.high then
```
Comment (jabraham17) [inline modules/standard/Sort.chpl:474]: same here, can you manually LICM the domain query?

**Review by jabraham17 — APPROVED** [review]

File: modules/standard/Sort.chpl (hunk 1 above)
Comment (lydia-duncan) [reply modules/standard/Sort.chpl:556]: I thought about doing that but it's used pretty late in the body of the function, so most of the body would still need to happen when `inPlaceAlgorithm=true`.  The other option would be to write a helper for the parts that will stay the same, but that's some fiddly code movement for something we don't want to live in a helper long term.

File: modules/standard/Sort.chpl (hunk 2 above)
Comment (lydia-duncan) [reply modules/standard/Sort.chpl:654]: You know, that's fair XD

File: modules/standard/Sort.chpl (hunk 1 above)
Comment (jabraham17) [reply modules/standard/Sort.chpl:556]: Sounds good, its not that much code duplication and its temporary anyways

File: modules/standard/Sort.chpl (hunk 4 above)
Comment (lydia-duncan) [reply modules/standard/Sort.chpl:613]: There's not a loop, but a local seems fine :)

File: modules/standard/Sort.chpl (hunk 5 above)
Comment (lydia-duncan) [reply modules/standard/Sort.chpl:474]: There's not a loop in this part of the operation but a local seems fine

**Pull request merged**